Homework #5

BolzmanBrain opened the pull request:
- developed CRUD Library Shell app

BolzmanBrain committed:
Homework #5

- developed CRUD Library Shell app

kataus commented on homework5/src/main/java/ru/otus/spring/dao/BookDaoH2.java:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    @Override
+    public Book createBookFromDto(BookDto dto) {
```
Это не задача dao слоя. Dao - это Data Access Object, а не маппинг. Он не должен знать деталей о слое предстаавления

BolzmanBrain replied:
Избавился от BookDto

kataus commented on homework5/src/main/java/ru/otus/spring/dao/BookDaoH2.java:
```diff
@@ -0,0 +1,134 @@
+        Author author;
+        Genre genre;
+        try {
+            author = authorDao.getById(dto.getIdAuthor()).orElseThrow();
```
Будет N+1 проблема

BolzmanBrain replied:
Исправил. Теперь использую join

kataus commented on homework5/src/main/java/ru/otus/spring/presentation/GenreToStringConvertorAsObject.java:
```diff
@@ -0,0 +1,12 @@
+public class GenreToStringConvertorAsObject implements GenreToStringConvertor {
+    @Override
+    public String convertToString(Genre genre) {
+        return genre.toString();
```
Настолько банальная задача, что стоит ли ради неё заводить отдельный слой?

BolzmanBrain replied:
Убрал эти классы.
Просто я их добавил из-за того что в ходе выполнения предыдущих домашек часто фигурировал вопрос: "А что, если нам потребуется изменить способ отображения объектов?"

kataus commented on homework5/src/main/java/ru/otus/spring/exceptions/ForeignKeyViolatedException.java:
```diff
@@ -0,0 +1,11 @@
+package ru.otus.spring.exceptions;
+
+public class ForeignKeyViolatedException extends Exception {
```
Принято максимально быстро уходить от checkException к unchecked. Это и удобнее и упрощает разработку. Первоисточник мысли : https://docs.oracle.com/javase/tutorial/essential/exceptions/runtime.html

BolzmanBrain replied:
Сделал свои эксепшены райнтаймовыми

kataus commented on homework5/src/main/java/ru/otus/spring/exceptions/UniqueKeyViolatedException.java:
```diff
@@ -0,0 +1,11 @@
+package ru.otus.spring.exceptions;
+
+public class UniqueKeyViolatedException extends Exception {
```
Удобно использовать своё единое корневое исключение. Тогда при обработке можно будет легко отделить наши и не наши исключения

BolzmanBrain replied:
Избавился от BookDto

kataus commented on homework5/src/main/java/ru/otus/spring/dao/dto/BookDto.java:
```diff
@@ -0,0 +1,15 @@
+public class BookDto {
+    private final Integer idBook;
+    private final String name;
+    private final Integer idAuthor;
```
У нас всё таки люди потребители продукта. АВторов и жанры им удобнее видеть в виде текста, а не ID

BolzmanBrain replied:
Избавился от BookDto

kataus commented on homework5/src/main/java/ru/otus/spring/domain/Book.java:
```diff
@@ -0,0 +1,11 @@
+import lombok.Data;
+
+@Data
+public class Book {
```
Похоже есть путаница в dto и book, вот тут ID нужны, это всё таки модель даннных, а не представления

BolzmanBrain replied:
Избавился от BookDto

BolzmanBrain committed:
Homework #5

- fixes after code review
- removed presentation package
- made my exeptions runtime
- removed BookDto
